influxdb3_write/src/persister.rs

```diff
@@ -74,8 +75,43 @@ impl Persister for PersisterImpl {
         }
     }
 
-    async fn load_segments(&self, _most_recent_n: usize) -> crate::Result<Vec<PersistedSegment>> {
-        todo!()
+    async fn load_segments(&self, most_recent_n: usize) -> crate::Result<Vec<PersistedSegment>> {
+        let segment_list = self
+            .object_store
+            .list(Some(&SegmentInfoFilePath::dir()))
+            .await?
+            .collect::<Vec<_>>()
+            .await;
+
+        // Why not collect into a Result<Vec<ObjectMeta>, object_store::Error>>
+        // like we could with Iterators? Well because it's a stream it ends up
+        // using different traits and can't really do that. So we need to loop
+        // through to return any errors that might have occurred, then do an
+        // unstable sort (which is faster and we know won't have any
+        // duplicates) since these can arrive out of order, and then issue gets
+        // on the n most recent segments that we want and is returned in order
+        // of the moste recent to least.
+        let mut list = Vec::new();
+        for segment in segment_list {
+            list.push(segment?);
+        }
+
+        list.sort_unstable_by(|a, b| a.location.cmp(&b.location));
+
+        let len = list.len();
+        let range = if len <= most_recent_n {
+            0..len
+        } else {
+            0..most_recent_n
+        };
+
+        let mut output = Vec::new();
+        for item in &list[range] {
+            let bytes = self.object_store.get(&item.location).await?.bytes().await?;
```

Comment:
I took a good stab at this, but because of async_trait hit a real fun "higher ranked order error". It should have worked, but the lifetimes are weird given async_trait and if we don't use async_trait we hit other errors. We could merge as is and open up an issue with my findings for follow up later to take a better look at this in the future and possibly make it work better. I also was unable to get it working with rayon either for parallelization.

Reply:
Let's merge as is and log an issue for grabbing the segments in parallel for the future. Better to at least get things working for the moment 😄

Reply:
Okay I'll do that after we merge this.